test/_lib/user-permissions.js

```diff
@@ -0,0 +1,219 @@
+  var method = task.method;
+  server.exchange.security[method](username, permissions, callback);
+}, 1);
+
+module.exports.allowMethod = function(server, username, component, method) {
+  var path = '/DOMAIN_NAME/' + component + '/' + method;
+  var permissions = { methods: {} };
+  permissions.methods[path] = { authorized: true };
+
+  return new Promise(function(resolve, reject) {
+    queue.push(
+      {
+        server: server,
+        username,
+        permissions: permissions,
+        method: 'addUserPermissions'
+      },
+      function(err) {
+        if (err) return reject(err);
+        resolve();
+      }
+    );
+  });
+};
+
+module.exports.denyMethod = function(server, username, component, method) {
+  var path = '/DOMAIN_NAME/' + component + '/' + method;
+  var permissions = { methods: {} };
+  permissions.methods[path] = {};
+
+  return new Promise(function(resolve, reject) {
+    queue.push(
+      {
+        server: server,
+        username,
+        permissions: permissions,
+        method: 'removeUserPermissions'
+      },
+      function(err) {
+        if (err) return reject(err);
+        resolve();
+      }
+    );
+  });
+};
+
+module.exports.allowWebMethod = function(server, username, path) {
+  var permissions = { web: {} };
+  permissions.web[path] = {
+    actions: ['get', 'put', 'post'],
+    description: 'a test web permission'
+  };
+  return new Promise(function(resolve, reject) {
+    queue.push(
+      {
+        server: server,
+        username,
+        permissions: permissions,
+        method: 'addUserPermissions'
+      },
+      function(err) {
+        if (err) return reject(err);
+        resolve();
+      }
+    );
+  });
+};
+
+module.exports.denyWebMethod = function(server, username, path) {
+  var permissions = { web: {} };
+  permissions.web[path] = {};
+
+  return new Promise(function(resolve, reject) {
+    queue.push(
+      {
+        server: server,
+        username,
+        permissions: permissions,
+        method: 'removeUserPermissions'
+      },
+      function(err) {
+        if (err) return reject(err);
+        resolve();
+      }
+    );
+  });
+};
+
+module.exports.allowEvent = function(server, username, component, event) {
+  var path = '/DOMAIN_NAME/' + component + '/' + event;
+  var permissions = { events: {} };
+  permissions.events[path] = { authorized: true };
+
+  return new Promise(function(resolve, reject) {
+    queue.push(
+      {
+        server: server,
+        username,
+        permissions: permissions,
+        method: 'addUserPermissions'
+      },
+      function(err) {
+        if (err) return reject(err);
+        resolve();
+      }
+    );
+  });
+};
+
+module.exports.denyEvent = function(server, username, component, event) {
+  var path = '/DOMAIN_NAME/' + component + '/' + event;
+  var permissions = { events: {} };
+  permissions.events[path] = {};
+
+  return new Promise(function(resolve, reject) {
+    queue.push(
```

Comment:
Does queue.push() not already return a Promise? I think in async@3 it does, pls check.